frame/chaos/src/lib.rs

```diff
@@ -0,0 +1,501 @@
+			TotalShares::<T>::try_mutate(|x| {
+				x.checked_add(amount.saturated_into::<u128>().into())
+					.ok_or(ArithmeticError::Overflow)
+			})?;
+
+			// Trigger event
+			Self::deposit_event(Event::<T>::Staked {
+				who: from.clone(),
+				stake: amount,
+				nft: instance_id,
+			});
+
+			Ok(instance_id)
+		}
+
+		fn unstake(to: &Self::AccountId, instance_id: &Self::InstanceId) -> DispatchResult {
+			// Make sure we execute a final claim before unstaking.
+			<Self as ChaosProtocol>::claim(to, instance_id)?;
+
+			let nft = T::get_protocol_nft::<ChaosStakingNFTOf<T>>(instance_id)?;
+
+			let now = T::Time::now();
+
+			// Possibly penalize the staked asset, Chaos in this case.
+			let (penalized_stake, penalty_amount) =
+				nft.penalize_early_unstake_amount(now.as_secs(), EarlyUnstakePenalty::<T>::get())?;
+
+			// Transfer back the (possibly penalized) staked amount.
+			let chaos_account = Self::account_id();
+			// NOTE(hussein-aitlahcen): no need to keep protocol account alive.
+			T::ChaosAsset::transfer(&chaos_account, &to, penalized_stake, false)?;
+
+			// Decrement total shares.
+			// TODO: decrement by penalized only????? the protocol would own this fraction
+			// indefinitely
+			TotalShares::<T>::try_mutate(|x| {
+				x.checked_sub(nft.stake.saturated_into::<u128>().into())
+					.ok_or(ArithmeticError::Underflow)
+			})?;
+
+			// Actually burn the NFT from the storage.
+			T::burn_protocol_nft::<ChaosStakingNFTOf<T>>(instance_id)?;
+
+			// Trigger event
+			Self::deposit_event(Event::<T>::Unstaked {
+				to: to.clone(),
+				stake: nft.stake,
+				penalty: penalty_amount,
+				nft: *instance_id,
+			});
+
+			Ok(())
+		}
+
+		fn transfer_reward(
+			from: &Self::AccountId,
+			asset: Self::AssetId,
+			amount: Self::Balance,
+			keep_alive: bool,
+		) -> DispatchResult {
+			// Make sure the asset has been registered as reward, otherwise we don't track it.
+			ensure!(RewardAssets::<T>::get().contains(&asset), Error::<T>::RewardAssetDisabled);
+
+			// Transfer the reward locally.
+			let chaos_account = Self::account_id();
+			T::Assets::transfer(asset, from, &chaos_account, amount, keep_alive)?;
+
+			// Increment the reward index, used to compute user rewards.
+			RewardIndexes::<T>::try_mutate(asset, |x| {
+				let lifted_amount = RewardIndex::from(amount.saturated_into::<u128>());
+				match x {
+					Some(index) => {
+						*index =
+							(*index).checked_add(lifted_amount).ok_or(ArithmeticError::Overflow)?;
+					},
+					None => {
+						*x = Some(lifted_amount);
+					},
+				}
+				Ok(())
+			})
+		}
+
+		fn claim(to: &Self::AccountId, instance_id: &Self::InstanceId) -> DispatchResult {
+			T::try_mutate_protocol_nft(instance_id, |nft: &mut ChaosStakingNFTOf<T>| {
+				// NOTE(hussein-aitlahcen): user is still able to claim even if his position
+				// 'expired', do we allow that?
+
+				let share = U256::from(nft.stake.saturated_into::<u128>());
+				let total_shares = TotalShares::<T>::get();
```

Comment:
so if somebody adds share after me, he dillutes my share, also i was first adopter.

Reply:
and goal is to put for as short peruid as possible and take as fast as possible

Reply:
Updated with reward multiplier applied to the staked amount